### PR TITLE
Partial python 3 support

### PR DESCRIPTION
**Pull request by sirex**
This is part of bringing Python 3 support to CKAN: https://github.com/ckan/ckan/issues/3309

Things left:

- [x] https://github.com/benosteen/pairtree/pull/2
- [ ] don't know how to run tests with working `test.ini`
- [ ] boto s3 does not work with Python 3

I think this is good to merge, because at least package can be installed and imported under python 3. And that is a good start.

### CODE REVIEW

**Comment by sirex** [comment]
Ok, pairtree got python 3 support, and I did run tests, but only those which
not require external services.

It look, that this project is no longer maintained and I thing it is good to
merge, at least with partial python support. This will allow to move forward
with ckan python 3 migration. If there will be more python 3 incompatibility
issues found, they can be fixed later.

For now python package is installable and importable under python 3 and even
some tests pass. I think that is a good start.

So please, review this and release new version. Thanks.

Here are the test results:

    $ py.test --ignore=test/botostore_test.py\
              --ignore=test/swiftstore_test.py\
              --ignore=test/entry_points_test.py\
              --ignore=test/reststore_test.py\
              --cov=ofs\
              --cov-report=term-missing\
              test

    ============================= test session starts ==============================
    platform linux -- Python 3.6.4, pytest-3.4.1, py-1.5.2, pluggy-0.6.0
    rootdir: /home/sirex/devel/contrib/ckan/ofs, inifile:
    plugins: cov-2.5.1
    collected 23 items

    test/metadatastore_test.py .......                                       [ 30%]
    test/pairtree_test.py .......                                            [ 60%]
    test/zipofs_test.py .........                                            [100%]

    ----------- coverage: platform linux, python 3.6.4-final-0 -----------
    Name                         Stmts   Miss  Cover   Missing
    ----------------------------------------------------------
    ofs/__init__.py                  3      0   100%
    ofs/base.py                     26     11    58%   32, 39, 47, 54, 62, 71, 81, 86, 91, 98, 103
    ofs/command.py                 138    138     0%   1-202
    ofs/factory.py                   4      2    50%   4-5
    ofs/local/__init__.py            4      0   100%
    ofs/local/filestore.py          26     26     0%   1-42
    ofs/local/metadatastore.py      78     23    71%   63-64, 68-69, 87, 94-98, 101-107, 110-111, 114-115, 123-124
    ofs/local/pairtreestore.py     113     29    74%   36, 57, 61, 66-68, 71, 79, 96-97, 103, 111-115, 118-121, 128, 139, 150, 153-161
    ofs/local/storedjson.py         63     18    71%   8-9, 30-31, 40-42, 46-49, 58, 67-68, 73-74, 77-80
    ofs/local/zipfile.py           873    530    39%   17-19, 139-144, 151-160, 166-192, 202-250, 285, 290, 299, 323, 334-340, 354-357, 360-363, 367-399, 437, 440-444, 447-450, 454-459, 484, 496, 507-549, 553-558, 561, 576, 598, 605, 610-620, 649, 653-659, 683-687, 690, 693-705, 708, 711, 716-722, 726-785, 798, 802-805, 809-818, 831, 840, 842, 849, 856, 866, 871, 874, 883-904, 914-920, 928-932, 941-969, 975, 977, 979, 982, 985, 988-989, 991-992, 997-1073, 1086, 1089, 1093, 1101-1104, 1113, 1121-1146, 1169-1172, 1178-1179, 1186-1191, 1206-1218, 1233-1245, 1249-1252, 1281-1332, 1341-1362, 1366-1435, 1438
    ofs/local/zipstore.py          197     73    63%   23-24, 42, 44, 52, 59-65, 73-85, 93, 127, 142, 152-157, 166, 170, 174, 183-188, 199, 243-246, 260-263, 271-275, 279-281, 295-296, 298-300, 307, 312-325
    ofs/remote/__init__.py           0      0   100%
    ofs/remote/botostore.py        134    134     0%   7-226
    ofs/remote/proxystore.py        81     81     0%   1-113
    ofs/remote/reststore.py        135    135     0%   1-169
    ofs/remote/swiftstore.py       114    114     0%   5-169
    ----------------------------------------------------------
    TOTAL                         1989   1314    34%


    ========================== 23 passed in 0.25 seconds ===========================

Python 2 works too:

    $ py.test --ignore=test/botostore_test.py\
              --ignore=test/swiftstore_test.py\
              --ignore=test/entry_points_test.py\
              --ignore=test/reststore_test.py\
              --cov=ofs\
              --cov-report=term-missing\
              test

    ========================== test session starts =================================
    platform linux2 -- Python 2.7.14, pytest-3.4.1, py-1.5.2, pluggy-0.6.0
    rootdir: /home/sirex/devel/contrib/ckan/ofs, inifile:
    plugins: cov-2.5.1
    collected 23 items

    test/metadatastore_test.py .......                                        [ 30%]
    test/pairtree_test.py .......                                             [ 60%]
    test/zipofs_test.py .........                                             [100%]

    ---------- coverage: platform linux2, python 2.7.14-final-0 ----------
    Name                         Stmts   Miss  Cover   Missing
    ----------------------------------------------------------
    ofs/__init__.py                  3      0   100%
    ofs/base.py                     26     11    58%   32, 39, 47, 54, 62, 71, 81, 86, 91, 98, 103
    ofs/command.py                 138    138     0%   1-202
    ofs/factory.py                   4      2    50%   4-5
    ofs/local/__init__.py            4      0   100%
    ofs/local/filestore.py          26     26     0%   1-42
    ofs/local/metadatastore.py      78     23    71%   63-64, 68-69, 87, 94-98, 101-107, 110-111, 114-115, 123-124
    ofs/local/pairtreestore.py     113     29    74%   36, 57, 61, 66-68, 71, 79, 96-97, 103, 111-115, 118-121, 128, 139, 150, 153-
    161
    ofs/local/storedjson.py         63     18    71%   8-9, 30-31, 40-42, 46-49, 58, 67-68, 73-74, 77-80
    ofs/local/zipfile.py           873    529    39%   17-19, 139-144, 151-160, 166-192, 202-250, 285, 290, 299, 323, 334-340, 354-355, 360-363, 367-399, 437, 440-444, 447-450, 454-459, 484, 496, 507-549, 553-558, 561, 576, 598, 605, 610-620, 649, 653-659, 683-687, 690, 693-705, 708, 711, 716-722, 726-785, 798, 802-805, 809-818, 831, 840, 842, 849, 856, 866, 871, 874, 883-904, 914-920, 928-932, 941-969, 975, 977, 979, 982, 985, 988-989, 991-992, 997-1073, 1086, 1089, 1093, 1101-1104, 1113, 1121-1146, 1169-1172, 1178-1179, 1186-1191, 1206-1218, 1233-1245, 1249-1252, 1281-1332, 1341-1362, 1366-1435, 1438
    ofs/local/zipstore.py          197     71    64%   23-24, 42, 44, 52, 59-65, 73-85, 93, 127, 152-157, 166, 170, 174, 183-188, 199, 243-246, 260-263, 271-275, 279-281, 298-300, 302, 307, 312-325
    ofs/remote/__init__.py           0      0   100%
    ofs/remote/botostore.py        134    134     0%   7-226
    ofs/remote/proxystore.py        81     81     0%   1-113
    ofs/remote/reststore.py        135    135     0%   1-169
    ofs/remote/swiftstore.py       114    114     0%   5-169
    ----------------------------------------------------------
    TOTAL                         1989   1311    34%

**Comment by amercader** [comment]
Thanks @sirex.
As you say this is no longer maintained and we are very likely going to drop the `ofs` requirement in CKAN soon but it's welcomed anyway

**Comment by sirex** [comment]
Thanks @amercader, I know about plans to remove `ofs`, but it looked safer and easier to just make `ofs` installable and importable on python3, than removing it from `ckan`.

If you could also release new `ofs` version to pypi, it would be perfect!

**Comment by amercader** [comment]
I don't have permissions on pypi. @rufuspollock looks like you have, do you mind adding me as maintainer?

File: ofs/local/zipstore.py [hunk 1]
```diff
@@ -40,28 +41,28 @@ def __init__(self, zipfile, mode="r", compression=ZIP_STORED, allowZip64=False,
...
+            print("IMPORTANT: You MUST .close() this ZOFS instance for it to write the ending records in '%s' mode. Otherwise the resultant zip archive will be unreadable." % mode)
         self.zipfile = zipfile
         self.mode = mode
         self.compression = compression
         self.allowZip64 = allowZip64
         self.hashing_type = hashing_type
         self.quiet = quiet
         if mode == "r" and not is_zipfile(zipfile):
-            raise BadZipArchive, e
+            raise e
```
Comment (cclauss) [inline ofs/local/zipstore.py]: __e__ is an undefined name in this context...  Perhaps __raise BadZipArchive(zipfile + ' is not a valid zipfile.')__

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./ofs/local/zipstore.py:52:19: F821 undefined name 'e'
            raise e
                  ^
```

**Comment by rufuspollock** [comment]
@amercader added you - can you also make sure there is an `okfn` or `oki` user or similar on pypi and they have access too (to avoid this in future).

**Comment by amercader** [comment]
@sirex there's 0.4.3 now on pypi

**Pull request merged**